### PR TITLE
Start working on our base box

### PR DESCRIPTION
**Pull request by BronsonQuick**
This adds support for two "modes": normal and base.

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -52,8 +52,13 @@ Vagrant.configure("2") do |config|
 		end
 	end
 
-	# We <3 Ubuntu LTS
-	config.vm.box = "bento/ubuntu-16.04"
+	if CONF['_mode'] == "normal"
+		# Use the Chassis box we've built Php 7.0, MySQL 5.7 and nginx 1.10.3.
```
Comment (rmccue) [inline Vagrantfile]: I'd probably just say "with default config"

File: Vagrantfile [hunk 2]
```diff
@@ -100,7 +105,11 @@ Vagrant.configure("2") do |config|
 
 	# Help the user out the first time they provision
 	config.vm.provision :shell do |shell|
-		shell.path = "puppet/postprovision.sh"
+		if CONF['_mode'] == "base"
+			shell.path = "puppet/preparebox.sh"
+		else
+			shell.path = "puppet/postprovision.sh"
+		end
```
Comment (rmccue) [inline Vagrantfile]: So right now, `base` will run `preparebox`, which means it's only really useful for building the base box. Since we offer this as a config option, should we distinguish between the "building the base box" and the "just using the base box instead of a custom box" use cases?

File: config.yaml [hunk 3]
```diff
@@ -8,6 +8,14 @@
 # - project/config.local.yaml          (global overrides)
 # - project/config.yaml                 (global defaults)
 
+# By default we set the mode to normal.
+# This uses a Chassis box we've built that has Php 7.0, MySQL 5.7 and nginx 1.10.3.
+# Set to base if you'd like to configure Php versions, add extensions and make other modifications.
+# When you set this to base your initial provision will take longer as it will start with the bento/ubuntu-16.04 box.
```
Comment (rmccue) [inline config.yaml]: * Needs short description on first line.
* These should be line wrapped at 78-80 chars.
* Can we move this way to the bottom instead? We don't want people using it much.

Also, re: "set to base if ...", those can still be set, right?

I'd probably reword this to:
```
Base box mode

By default, we use a box built with the default settings. This speeds up
initial provisioning time and reduces disk usage by sharing the common parts of
the VM.

Set this to `base` to build your own base box.

Values: normal, base
```

**Comment by rmccue** [comment]
Also, can we add a script to handle the build -> publish to Vagrant Cloud process? Deploying a new version should be a single command.

**Comment by rmccue** [comment]
One blocker I've spotted is that we never finished https://github.com/rmccue/puppet-wp/pull/36 so it's not currently possible to provision multisite with this base box. We need to fix that.

Additionally, using a custom host is annoyingly too late, since it won't update the database with the new host. Not sure how to fix that. We could drop the database after provisioning the base box and let that run on first actual provision? That would also work around the previous issue.

**Comment by BronsonQuick** [comment]
>Also, can we add a script to handle the build -> publish to Vagrant Cloud process? Deploying a new version should be a single command.
I've done a first pass of that here: https://github.com/Chassis/Chassis/pull/476/files#diff-802e8cff0cd0d29b5c288195859ca853 Let me know any changes and/or enhancements you'd like me to add.

>One blocker I've spotted is that we never finished rmccue/puppet-wp#36 so it's not currently possible to provision multisite with this base box. We need to fix that.

>Additionally, using a custom host is annoyingly too late, since it won't update the database with the new host. Not sure how to fix that. We could drop the database after provisioning the base box and let that run on first actual provision? That would also work around the previous issue.

Hopefully https://github.com/rmccue/puppet-wp/pull/43 helps us get passed this blocker or at least helps us progress forward with it.

**Comment by BronsonQuick** [comment]
@rmccue I've finally got around to adding some more updates to this branch. Can you please have another look over it and offer some new feedback please Sir.

I know I will have to look into https://github.com/Chassis/Chassis/issues/601 and https://github.com/Chassis/Chassis/issues/582 as they will be issues if this is merged as is but I'll fix that up once I've got your new feedback on this. Taa!

**Comment by BronsonQuick** [comment]
@rmccue Ping for another review of this too please! 🙇‍♂️

**Comment by BronsonQuick** [comment]
@rmccue @kadamwhite And another one that I'm waiting on another review for. Taa muchly!

File: puppet/manifests/development.pp [hunk 4]
```diff
@@ -19,6 +19,12 @@
 	}
 }
 
+class { 'apt':
+	update => {
+		frequency => 'weekly',
```
[the remark this answers is not in the document]
Comment (roborourke) [reply puppet/manifests/development.pp]: Oh yeah - my bad, I didn’t explain very well, was thinking if there’s somewhere we could do that if say the config has been modified or the php version was changed perhaps. Not something to do here I don’t think.

File: Vagrantfile [hunk 5]
```diff
@@ -157,7 +162,11 @@ Vagrant.configure("2") do |config|
 
 	# Help the user out the first time they provision
 	config.vm.provision :shell do |shell|
-		shell.path = "puppet/postprovision.sh"
+		if CONF['_mode'] == "base"
```
[the remark this answers is not in the document]
Comment (roborourke) [reply Vagrantfile]: I was confused as the preprovision script is empty, I thought that’d be the one to run on the custom base box. I still don’t really follow this but all good!

[the remark this answers is not in the document]
Comment (BronsonQuick) [reply Vagrantfile]: To be honest I'm a bit confused about the preprovision script being empty as well. It was something Ryan suggested when he had a look over the initial PR when we were in the office. @rmccue can you remember your logic for adding that by any chance?
